cpp/ql/src/semmle/code/cpp/ir/implementation/IRType.qll

```diff
@@ -139,34 +139,41 @@ class IRNumericType extends IRSizedType {
   }
 }
 
+/**
+ * An integer type. This includes `IRSignedIntegerType` and `IRUnsignedIntegerType`.
+ */
+class IRIntegerType extends IRNumericType {
+  IRIntegerType() {
+    this = TIRSignedIntegerType(byteSize) or
+    this = TIRUnsignedIntegerType(byteSize)
+  }
+
+  pragma[noinline]
+  final override int getByteSize() { result = byteSize }
+}
+
 /**
  * A signed two's-complement integer. Also used to represent enums whose underlying type is a signed
  * integer, as well as character types whose representation is signed.
  */
-class IRSignedIntegerType extends IRNumericType, TIRSignedIntegerType {
+class IRSignedIntegerType extends IRIntegerType, TIRSignedIntegerType {
   final override string toString() { result = "int" + byteSize.toString() }
 
   final override Language::LanguageType getCanonicalLanguageType() {
     result = Language::getCanonicalSignedIntegerType(byteSize)
   }
-
-  pragma[noinline]
-  final override int getByteSize() { result = byteSize }
```

Comment:
This override was deliberately copied to each leaf class in https://github.com/github/codeql/pull/2272. While there might be no existing code in our repo that's sensitive to its position in the numeric-type hierarchy, I can easily imagine such code being written in the future.

My first thought was "why not move the override all the way up to `IRSizedType`", but then I remembered. If we want to keep the override duplicated at the leaves, there probably needs to be a code comment on each of them to alert future maintainers that it's deliberate.

Reply:
Fixed, and added comments in all the tempting places.